### PR TITLE
Add lru

### PR DESCRIPTION
**Pull request by pbougue**
JIRA: http://jira.canaltp.fr/browse/NAVP-480

### CODE REVIEW

File: source/routing/next_stop_time.cpp [hunk 1]
```diff
@@ -413,42 +413,62 @@ static void fill_cache(const DateTime from,
...
+    } else if (rt_level != other.rt_level) {
+        return rt_level < other.rt_level;
+    }
+    return accessibilite_params < other.accessibilite_params;
+}
+
+CachedNextStopTime CachedNextStopTimeManager::Fun::operator()(const CachedNextStopTimeKey& key) const {
+    CachedNextStopTime result;
+    auto logger = log4cplus::Logger::getInstance("log");
+    LOG4CPLUS_INFO(logger, "Cache miss : " << ++cache_miss);
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: nop a fan of the instance variable incremented in the log.

If we change the log level we might change the value in the variable :wink:

File: source/routing/next_stop_time.h [hunk 2]
```diff
@@ -197,12 +199,28 @@ struct NextStopTime {
     const type::Data& data;
 };
 
+struct CachedNextStopTimeKey {
+    uint32_t from; //first day concerned by the cache
```
Comment (antoine-de) [inline source/routing/next_stop_time.h]: insn't it a DateTime ?

File: source/routing/next_stop_time.h [hunk 3]
```diff
@@ -197,12 +199,28 @@ struct NextStopTime {
...
+    using DtSt = std::pair<DateTime, const type::StopTime*>;
+    IdxMap<JourneyPatternPoint, std::vector<DtSt>> departure;
+    IdxMap<JourneyPatternPoint, std::vector<DtSt>> arrival;
+};
 
-    void load(const type::Data& d,
-              const DateTime from,
-              const DateTime to,
+struct CachedNextStopTimeManager {
+    explicit CachedNextStopTimeManager(const type::Data& d, size_t miss = 0, size_t max_cache = 10) :
```
Comment (antoine-de) [inline source/routing/next_stop_time.h]: why do you pass the miss in param ?

File: source/routing/next_stop_time.h [hunk 4]
```diff
@@ -216,9 +234,17 @@ struct CachedNextStopTime {
...
 
 private:
-    using DtSt = std::pair<DateTime, const type::StopTime*>;
-    IdxMap<JourneyPatternPoint, std::vector<DtSt>> departure;
-    IdxMap<JourneyPatternPoint, std::vector<DtSt>> arrival;
+    struct Fun {
+        typedef CachedNextStopTimeKey const& argument_type;
+        typedef CachedNextStopTime result_type;
+        const type::Data& data;
+        mutable size_t cache_miss;
```
Comment (antoine-de) [inline source/routing/next_stop_time.h]: I don't understand you use of cache_miss ?

For me it should be the lru that handle this no ?

File: source/routing/next_stop_time.h (hunk 2 above)
Comment (pbougue) [inline source/routing/next_stop_time.h]: Initially it is, but it's converted to a Date in  next_stop_time.cpp:457.
Sounded the easiest way to compare without misleading (day-and so cache index- integer is nice).
And thinking of an hypothetical smaller cache time-window is a too big dev at least for now.

File: source/routing/next_stop_time.h [hunk 5]
```diff
@@ -216,9 +234,17 @@ struct CachedNextStopTime {
                    const bool clockwise) const;
 
 private:
-    using DtSt = std::pair<DateTime, const type::StopTime*>;
-    IdxMap<JourneyPatternPoint, std::vector<DtSt>> departure;
-    IdxMap<JourneyPatternPoint, std::vector<DtSt>> arrival;
+    struct Fun {
```
Comment (antoine-de) [inline source/routing/next_stop_time.h]: I think it deserve a better name :wink: 

CacheCreator ? (not sure I correctly understand it's use though)

File: source/routing/next_stop_time.cpp (hunk 1 above)
Comment (pbougue) [inline source/routing/next_stop_time.cpp]: Yep, will change that with the miss refactor.

File: source/routing/next_stop_time.h (hunk 3 above)
Comment (pbougue) [inline source/routing/next_stop_time.h]: Not sure where to cumulate that...

File: source/routing/next_stop_time.h (hunk 5 above)
Comment (pbougue) [inline source/routing/next_stop_time.h]: Yep, CacheCreator is a better name, will change that.

File: source/routing/next_stop_time.cpp [hunk 6]
```diff
@@ -413,42 +413,62 @@ static void fill_cache(const DateTime from,
...
         boost::sort(jpp_dtst.second, compare);
     }
+    return result;
+}
+
+void CachedNextStopTimeManager::load(const DateTime from,
+                                     const type::RTLevel rt_level,
+                                     const type::AccessibiliteParams& accessibilite_params) {
+    CachedNextStopTimeKey key(DateTimeUtils::date(from), rt_level, accessibilite_params);
+    cache = &lru(key);
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: not a fan of the variable changed by side effect.

I think the manager should return the cache (and the cache should be stored in Raptor no ?)

Comment (pbougue) [inline source/routing/next_stop_time.cpp]: Ok, can move it in the RAPTOR

File: source/routing/next_stop_time.h (hunk 4 above)
Comment (pbougue) [inline source/routing/next_stop_time.h]: To be seen in cache_miss tracing refactor

File: source/routing/next_stop_time.cpp [hunk 7]
```diff
@@ -413,31 +413,40 @@ static void fill_cache(const DateTime from,
...
+CachedNextStopTime CachedNextStopTimeManager::CacheCreator::operator()(const CachedNextStopTimeKey& key) const {
+    CachedNextStopTime result;
     const auto& jp_container = data.dataRaptor->jp_container;
 
-    departure.assign(jp_container.get_jpps_values());
-    arrival.assign(jp_container.get_jpps_values());
+    result.departure.assign(jp_container.get_jpps_values());
+    result.arrival.assign(jp_container.get_jpps_values());
+    DateTime dt_from = DateTimeUtils::set(key.from, 0);
+    DateTime dt_to = DateTimeUtils::set(key.from + 2, 0);
```
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: I think  a comment explaining the `+2` would be nice

Comment (xlqian) [inline source/routing/next_stop_time.cpp]: Would it be interesting that `2` is configurable?

File: source/routing/raptor.h [hunk 8]
```diff
@@ -70,7 +70,8 @@ struct RAPTOR
 {
     const navitia::type::Data& data;
 
-    CachedNextStopTime cached_next_st;
+    CachedNextStopTimeManager cached_next_st_manager;
+    const CachedNextStopTime* current_cache;
```
Comment (antoine-de) [inline source/routing/raptor.h]: you should init it with nullptr

I prefer the previous name for this (`next_st`), I think the cache thing only concern the CachedNextStopTimeManager, not the thing cached

File: source/routing/next_stop_time.cpp (hunk 7 above)
Comment (antoine-de) [inline source/routing/next_stop_time.cpp]: I don't think so, I can't see an instance where we'll want a different value

File: source/routing/next_stop_time.cpp [hunk 9]
```diff
@@ -465,7 +477,18 @@ CachedNextStopTime::next_stop_time(const StopEvent stop_event,
     return {nullptr, 0};
 }
 
+CachedNextStopTimeManager::~CachedNextStopTimeManager() {
+    auto logger = log4cplus::Logger::getInstance("log");
+    LOG4CPLUS_INFO(logger, "Cache miss : " << lru.get_nb_cache_miss() << " / " << lru.get_nb_calls());
```
Comment (xlqian) [inline source/routing/next_stop_time.cpp:482]: :+1:

File: source/routing/next_stop_time.h [hunk 10]
```diff
@@ -214,11 +225,28 @@ struct CachedNextStopTime {
                    const JppIdx jpp_idx,
                    const DateTime dt,
                    const bool clockwise) const;
+};
+
+struct CachedNextStopTimeManager {
+    explicit CachedNextStopTimeManager(const type::Data& d, size_t max_cache = 10) :
```
Comment (xlqian) [inline source/routing/next_stop_time.h]: `d` ... can you rename it :)

File: source/routing/next_stop_time.h [hunk 11]
```diff
@@ -214,11 +225,28 @@ struct CachedNextStopTime {
...
+    const CachedNextStopTime* load(const DateTime from,
+                                   const type::RTLevel rt_level,
+                                   const type::AccessibiliteParams& accessibilite_params);
 
 private:
-    using DtSt = std::pair<DateTime, const type::StopTime*>;
-    IdxMap<JourneyPatternPoint, std::vector<DtSt>> departure;
-    IdxMap<JourneyPatternPoint, std::vector<DtSt>> arrival;
+    struct CacheCreator {
+        typedef CachedNextStopTimeKey const& argument_type;
```
Comment (xlqian) [inline source/routing/next_stop_time.h:243]: `argument_type` is used?

Comment (TeXitoi) [inline source/routing/next_stop_time.h:243]: Yes by Lru

**Pull request merged**